include/rmm/mr/device/binning_memory_resource.hpp

```diff
@@ -0,0 +1,190 @@
+
+#include <algorithm>
+#include <cassert>
+#include <map>
+#include <memory>
+#include <vector>
+
+namespace rmm {
+namespace mr {
+
+/**
+ * @brief Allocates memory from upstream resources associated with bin sizes.
+ *
+ * @tparam UpstreamResource memory_resource to use for allocations that don't fall within any
+ * configured bin size. Implements rmm::mr::device_memory_resource interface.
+ */
+template <typename Upstream>
+class binning_memory_resource final : public device_memory_resource {
+ public:
+  // The required alignment of this allocator
+  static constexpr std::size_t allocation_alignment = 256;
+
+  /**
+   * @brief Construct a new binning memory resource object
+   *
+   * Initially has no bins, so simply uses the upstream_resource until bin resources are added
+   * with `add_bin`.
+   *
+   * @throws rmm::logic_error if size_base is not a power of two.
+   *
+   * @param upstream_resource The upstream memory resource used to allocate pools of blocks
+   */
+  explicit binning_memory_resource(Upstream* upstream_resource) : upstream_mr_{upstream_resource} {}
+
+  /**
+   * @brief Destroy the binning_memory_resource and free all memory allocated from the upstream
+   * resource.
+   */
+  ~binning_memory_resource() = default;
+
+  binning_memory_resource()                               = delete;
+  binning_memory_resource(binning_memory_resource const&) = delete;
+  binning_memory_resource(binning_memory_resource&&)      = delete;
+  binning_memory_resource& operator=(binning_memory_resource const&) = delete;
+  binning_memory_resource& operator=(binning_memory_resource&&) = delete;
+
+  /**
+   * @brief Query whether the resource supports use of non-null streams for
+   * allocation/deallocation.
+   *
+   * @returns true
+   */
+  bool supports_streams() const noexcept override { return true; }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the resource supports get_mem_info, false otherwise.
+   */
+  bool supports_get_mem_info() const noexcept override { return false; }
+
+  /**
+   * @brief Get the upstream memory_resource object.
+   *
+   * @return UpstreamResource* the upstream memory resource.
+   */
+  Upstream* get_upstream() const noexcept { return upstream_mr_; }
+
+  /**
+   * @brief Add a bin allocator to this resource
+   *
+   * Adds `bin_resource` if it is not null; otherwise constructs and adds a
+   * fixed_size_memory_resource.
+   *
+   * This bin will be used for any allocation smaller than `allocation_size` that is larger than
+   * the next smaller bin's allocation size.
+   *
+   * If there is already a bin of the specified size nothing is changed.
+   *
+   * This function is not thread safe.
+   *
+   * @param allocation_size The maximum size that this bin allocates
+   * @param bin_resource The memory resource for the bin
+   */
+  void add_bin(std::size_t allocation_size, device_memory_resource* bin_resource = nullptr)
```

Comment:
Would it make sense to add a ctor that pre-populates bins based on a geometric growth like the previous resource had?

Reply:
I like the `add_bin` interface. The only potential downside is that it won't really work with the `owning_wrapper` if `bin_resource` is not `nullptr`. `owning_wrapper` is designed to work such that the upstreams are passed as ctor arguments. Not sure if you consider that to be a problem or not.

Reply:
Yeah I could add one.  

Also, owning wrapper will work if you pass owning wrappers as the resources, right?

Reply:
> Also, owning wrapper will work if you pass owning wrappers as the resources, right?

Since the `bin_resource` is passed as a `device_memory_resource*` there's nothing to keep the `shared_ptr` from the `owning_wrapper` alive.

Reply:
Why is it called owning wrapper if keeping it alive doesn't keep its owned MR alive?

Reply:
@jrhemstad I added this. @shwina @kkraus14 you may want to take a look at the Python changes in the latest commit. Simplifies usage a bit for a likely common case.

Reply:
> Why is it called owning wrapper if keeping it alive doesn't keep its owned MR alive?

It's nothing to do with `owning_wrapper`. It's the `add_bin` function. 

For example:
```
auto cuda = std::make_shared<cuda_memory_resource>();
auto pool = make_owning_wrapper<pool_memory_resource>(cuda); // pool keeps `cuda` alive

auto binning = make_owning_wrapper<binning_memory_resource>(pool); // binning keeps `pool` alive

auto fixed = make_owning_wrapper<fixed_memory_resource>(cuda, 1MiB); 

binning.add_bin(1MiB, fixed.get()); // binning will _not_ keep `fixed` alive 
```

Reply:
So are you suggesting to change it to `add_bin(size_t, std::shared_ptr<rmm::mr::device_memory_resource>)`?

Reply:
I'd either just live with the fact that `add_bin` is non-owning, or make `add_bin` a private member and only have constructors like the geometric growth one where you pass the upstream at construction.

Reply:
I prefer to live with it.